### ISSUE TITLE
Awaiting issues with grow/weaken

### ISSUE THREAD

**Issue by MatthewTh0** [danielyxie/bitburner#2904]
Version Bitburner v1.4.0 (49de4a28)

So my script below is called with `ns.run(early-hack-template.js, 3, targetServer);` (with the target server being of many servers such as rothman-uni, harakiri-sushi, iron-gym, netlink, sigma-cosmetics, and others).

- early-hack-template.js
```
/** @param {NS} ns **/
export async function main(ns) {
    const args = ns.flags([
        ["help", false],
        ["quitThresh", 1000],
        ["moneyThresh", 10000000]
    ]);

    if (args.help) {
        ns.tprint("Argument one is server to hack.");
        ns.tprint("Available flags are [--help], [--quitThresh moneyValue], and [--moneyThresh moneyValue].");
        return;
    }

    const defaultTarget = "foodnstuff",
        quitThresh = args.quitThresh;


    let target = defaultTarget,
        moneyThresh = args.moneyThresh, 
        moneyOnServer=0;

    // args[0] = target server
    // Defines the "target server", which is the server
    // that we're going to hack. In this case, the default is "foodnstuff"
    switch (ns.args.length) {
        default:
        case 1:
            target = ns.args[0];
        case 0:
            // uses default server if no args
            break;
    }

    // Defines how much money a server should have before we hack it
    // In this case, it is set to 75% of the server's max money
    // or can be set through the flag --moneyThresh


    if (moneyThresh == 10000000) {
        moneyThresh = ns.getServerMaxMoney(target) * 0.75;
    }

    // Defines the maximum security level the target server can
    // have. If the target's security level is higher than this,
    // we'll weaken it before doing anything else
    let securityThresh = ns.getServerMinSecurityLevel(target) + 5;

    // If needed and we have the necessary programs, 
    // use them to open ports on the target server
    if (!ns.hasRootAccess(target)) {
        if (ns.fileExists("BruteSSH.exe", "home")) {
            ns.brutessh(target);
        }
        // Get root access to target server
        ns.nuke(target);
        // If it now has root access, continue
        // otherwise stop running the script.
        if (ns.hasRootAccess(target)) {
            ns.tprint("Success: Hacked server " + target + "\nBut it is necessary to manually install backdoor on the server.");
        } else {
            ns.tprint("Failure: This script was unable to hack server: " + target + ". Try another hacking script or wait until more ports can be opened.");
            return;
        }
    }
    // Infinite loop that continuously hacks/grows/weakens the target server
    while (true) {
        moneyOnServer = ns.getServerMoneyAvailable(target);
        if (ns.getServerSecurityLevel(target) > securityThresh) {
            // If the server's security level is above our threshold, weaken it
            await ns.weaken(target);
        } else if (moneyOnServer < quitThresh) {
            ns.alert("Managed to essentially bankrupt the server: " + target + " Killing hack script...");
            return;
        } else if (moneyOnServer < moneyThresh) {
            // If the server's money is less than our threshold, grow it
            await ns.grow(target);
        } else {
            // Otherwise, hack it
            await ns.hack(target);
        }
    }
}
```

It then errors saying: 

- Error message example
```
RUNTIME ERROR
/scripts/hacking/early-hack-template.js@home
Args: ["harakiri-sushi"]

Concurrent calls to Netscript functions are not allowed! Did you forget to await hack(), grow(), or some other promise-returning function? (Currently running: grow tried to run: getServerMoneyAvailable())
```

Furthermore there's some weird things about this error. First, I have this script running fine on other servers that I copy it to, it only "crashes" on home. Secondly, the crash itself is weird, it often continues running despite the crash with logs such as these below. Often "crashing" every so often. It also shows the crash error when manually killed as well. 

- Part of a tail/log for early-hack-template.js rothman-uni 
```
getServerMaxMoney: returned $566.762m for 'rothman-uni'

getServerMinSecurityLevel: returned 16.000 for rothman-uni

getServerMoneyAvailable: returned $151.136m for 'rothman-uni'

getServerSecurityLevel: returned 48.000 for 'rothman-uni'

weaken: Executing on 'rothman-uni' in 25 minutes 6.946 seconds (t=4)

asleep: Sleeping for 200 milliseconds

asleep: Sleeping for 1000 milliseconds

Script crashed with runtime error

asleep: Sleeping for 200 milliseconds

asleep: Sleeping for 1000 milliseconds

Script crashed with runtime error

asleep: Sleeping for 200 milliseconds

asleep: Sleeping for 1000 milliseconds

Script crashed with runtime error
```

I am using a different variation now, but I've done nothing to the while loop besides throwing in some asleeps/sleeps after the last else or after the weaken/grow, which doesn't seem to make a difference somehow (as shown above, although I've haven't tried going above one second at any place so far). I've attached my latest save file where this type of script still causes errors. I'm about to finally bother switching my hacking methods (I've been focusing on literally everything else so far), but I'm still confused why this is happening. I posted this to the subreddit a while ago and got no real ideas besides trying to asleep which as I said didn't work. Online, all I could find is that javascript sleeps are typically kinda just wonky timeouts (that's not saying your implementation is bad, from what I understand just a drawback of javascript itself) and may be the underlying cause. I think it's coincedence, but most of times I've noticed it, the targets tend to have a hyphen in their name (but not always such as netlink).

[bitburnerSave_1643869050_BN4x0.zip](https://github.com/danielyxie/bitburner/files/7992279/bitburnerSave_1643869050_BN4x0.zip)

**Comment by MatthewTh0** [comment]
They are, and alert doesn't need await.

**Comment by zeddrak** [comment]
Hmm, it says "Currently running: grow tried to run: getServerMoneyAvailable()", which implies that it didn't await the grow properly before looping. Try adding an await ns.sleep(1) right before the `moneyOnServer = ns.getServerMoneyAvailable(target);`

**Comment by MatthewTh0** [comment]
So, I'm 99.99% sure I figured it out. The issue wasn't in the code you saw. The issue was that I was using ns as a global variable in order to get VSCode to autcomplete things like so 
``` javascript
/** @typedef {import("").NS} NS */
/** @type {NS} */
let ns = null;
export async function main(_ns) {
  ns = _ns;
```
I didn't show this as I thought it didn't matter and I hadn't seen any side-effects or at least I didn't notice that this error started shortly after I started doing this. So, when I was running the same script on the same server, it would share the ns object and "run concurrently" with each other.  I feel really stupid honestly since I think I did glimpse over this thread for a second which has a comment that describes the issue (that I obviously didn't take in fully). [r/Bitburner thread](https://www.reddit.com/r/Bitburner/comments/ry6wno/is_this_the_only_way_for_netscript_autocomplete/) 
I just need to remember to never use the global ns if the script might run multiple instances on the same server.
